packages/@aws-cdk/aws-iam/lib/grant.ts

```diff
@@ -0,0 +1,78 @@
+import { PolicyStatement } from "./policy-document";
+import { IPrincipal } from "./principals";
+
+/**
+ * Properties for a grant operation
+ */
+export interface GrantProps {
+  /**
+   * The principal to grant to
+   *
+   * @default No work is done
+   */
+  principal?: IPrincipal;
+
+  /**
+   * The actions to grant
+   */
+  actions: string[];
+
+  /**
+   * The resource ARNs to grant to
+   */
+  resourceArns: string[];
+
+  /**
+   * Adder to the resource policy
+   *
+   * @default Failure if no principal policy and `skipResourcePolicy` is not set.
+   */
+  addToResourcePolicy?: (statement: PolicyStatement) => void;
+
+  /**
+   * When referring to the resource in a resource policy, use this as ARN.
+   *
+   * (Depending on the resource type, this needs to be '*' in a resource policy).
+   *
+   * @default Same as regular resource ARNs
+   */
+  resourceSelfArns?: string[];
+
+  /**
+   * If there is no resource policy, ignore the error
+   *
+   * @default false
+   */
+  skipResourcePolicy?: boolean;
+}
+
+/**
+ * Helper function to implement grants.
+ *
+ * The pattern is the same every time. We try to add to the principal
+ * first, then add to the resource afterwards.
+ *
+ * @returns false if `skipResourcePolicy` was used, true otherwise
+ */
+export function grant(props: GrantProps): boolean {
```

Comment:
> I am okay with leaving it as is since it's basically a utility for L2s, 

That was my thinking behind this choice.

> an object that implements an IResourcePolicy interface with an addToResourcePolicy method just pass in the granting construct (i.e. Queue).

I suppose. Considered that as well. Actually TypeScript's structural typing might make that very easy.

> Also, do you think that maybe we should rename props to options (or opts) and GrantProps to GrantOptions (we said that Props are reserved to ctors and factory methods, no?)

I'm flip-flopping on this. On the one hand, makes perfect sense to me. On the other hand, I've seen people develop in Java and they need to type the class name every time, and I bet it'll be grating to them if it's "inconsistent" in that way.

Reply:
In Java the IDE will help, so I am not that concerned there